rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+visualise the node graph.
+
+Generating the node graph automatically based on a configuration/manifest file
+should simplify this. It should be easier to figure out what the node graph will
+look like through inspecting a manifest file (assuming it has an easily
+understandable format). Additional benefits of configuration-based node creation
+include:
+
+- It would be easier to build tools to visualise the node graph without running
+  the application.
+- It would be easier to build tools to perform static analysis of the node graph
+  to detect potential problems.
+- It would usually be easier to reason about the security of a static node
+  graph.
+
+## Overview
+
+Moving all node and channel creation to a configuration-based mechanism is a
+significant and potentially disruptive change. This RFC proposes a more gradual
+approach by implementing a re-usable Wasm-based planner node that uses a
+manifest file to do the node and channel creation. The functionality can be
+expanded over time to support more complex scenarios.
+
+The planner node will be responsible for creating nodes, creating channels,
+sending channel handles to the created nodes to ensure that the full node graph
+is created on start-up. It will also act as a message router, which would route
+HTTP- and GRPC invocations to other nodes based on the associated labels.
+
+Some applications require dynamic creation of nodes and channels based on the
+labels associated with incoming invocations. A sub-graph template mechanism will
+be implemented to support dynamic node creation. The typical scenario where this
+is required is the case where a per-user node must be created to support
+per-user labels. Although this means that the entire node graph cannot be
+created up-front and that the node graph will not be static, it should still be
+easier to reason about each of the sub-graphs and the overall node graph is
+still strongly constrained.
+
+## Detailed Design
+
+### Planner Node
+
+The planner node will be the initial node created for an application. The
+manifest file will be passed to the planner node as part of the application's
+`ConfigMap`.
+
+The planner node will create all the nodes and channels as specified with the
+appropriate labels. It will send initialisation messages to each of the nodes,
+providing additional configuration and handles to channels. It is the
+responsibility of each of the nodes to store these channels for future
+communications with other nodes. It will send a filtered version of the config
+map as part of the initialisation message, based on the configuration keys
+specified in the `config_files` field.
+
+### Node and Channel Creation
+
+Creation of static nodes and channels is straight-forward. The planner node will
+read these on start-up and use the values specified in the manifest to create
+the nodes and channels and send the read- and write handles to the appropriate
+nodes.
```

Comment:
I don't think we can get away with node creation not happening anywhere, see the PIR example. In fact, perhaps it would be good to have a section about that example in particular, and how it would be expressed via this planner node?